crates/astria-composer/src/searcher/bundle_factory.rs

```diff
@@ -0,0 +1,479 @@
+
+        if seq_action_size > self.max_size {
+            return Err(SizedBundleError::SequenceActionTooLarge(seq_action));
+        }
+
+        if self.curr_size + seq_action_size > self.max_size {
+            return Err(SizedBundleError::NotEnoughSpace(seq_action));
+        }
+
+        self.buffer.push(Action::Sequence(seq_action));
+        self.curr_size += seq_action_size;
+        Ok(())
+    }
+
+    /// Replace self with a new empty bundle, returning the old bundle.
+    fn flush(&mut self) -> SizedBundle {
+        mem::replace(self, Self::new(self.max_size))
+    }
+
+    /// Consume self and return the underlying buffer of actions.
+    fn into_actions(self) -> Vec<Action> {
+        self.buffer
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum BundleFactoryError {
+    #[error("sequence action is larger than the max bundle size. seq_action size: {size}")]
+    SequenceActionTooLarge { size: usize, max_size: usize },
+}
+
+/// Manages the bundling of sequence actions into `SizedBundle`s. A `Vec<Action>` is flushed and
+/// added to the `finished` queue when an incoming `SequenceAction` won't fit in the current bundle.
+/// The `finished` queue operates in FIFO order, where `Vec<Action>`s are added to the back and
+/// taken off from the front.
+pub(super) struct BundleFactory {
+    /// The current bundle being built.
+    curr_bundle: SizedBundle,
+    /// The queue of bundles that have been built but not yet sent to the sequencer.
+    finished: VecDeque<SizedBundle>,
+}
+
+impl BundleFactory {
+    pub(super) fn new(max_bytes_per_bundle: usize) -> Self {
+        Self {
+            curr_bundle: SizedBundle::new(max_bytes_per_bundle),
+            finished: VecDeque::new(),
+        }
+    }
+
+    /// Buffer `seq_action` into the current bundle. If the bundle won't fit `seq_action`, flush
+    /// `curr_bundle` into the `finished` queue and start a new bundle
+    pub(super) fn try_push(
+        &mut self,
+        seq_action: SequenceAction,
+    ) -> Result<(), BundleFactoryError> {
+        let seq_action_size = estimate_size_of_sequence_action(&seq_action);
+
+        match self.curr_bundle.push(seq_action) {
+            Err(SizedBundleError::SequenceActionTooLarge(_seq_action)) => {
+                // reject the sequence action if it is larger than the max bundle size
+                Err(BundleFactoryError::SequenceActionTooLarge {
+                    size: seq_action_size,
+                    max_size: self.curr_bundle.max_size,
+                })
+            }
+            Err(SizedBundleError::NotEnoughSpace(seq_action)) => {
+                // if the bundle is full, flush it and start a new one
+                self.finished.push_back(self.curr_bundle.flush());
+                self.curr_bundle
+                    .push(seq_action)
+                    .expect("seq_action should not be larger than max bundle size");
```

Comment:
Why is panicking here ok? Because of `SizedBundle::push`? Can you change the `expect` message to that effect?